### PR TITLE
feat(connector): [Airwallex] add authorize, capture, void, psync, Webhooks support 

### PR DESCRIPTION
**Pull request by SamraatBansal**
## Type of Change

- [ ] Bugfix
- [X] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description
Updating support for [Airwallex](https://www.airwallex.com/docs/api?v=2017-01-01#/Introduction) payment gateway
Payment Method: card (Non 3DS)
Supported payment flows

- [Authorize](https://www.airwallex.com/docs/api?v=2017-01-01#/Payment_Acceptance/Payment_Intents/Intro)
- [Capture](https://www.airwallex.com/docs/api?v=2017-01-01#/Payment_Acceptance/Payment_Intents/_api_v1_pa_payment_intents__id__capture/post)
- [PSync](https://www.airwallex.com/docs/api?v=2017-01-01#/Payment_Acceptance/Payment_Intents/_api_v1_pa_payment_intents__id_/get)
- [Cancel](https://www.airwallex.com/docs/api?v=2017-01-01#/Payment_Acceptance/Payment_Intents/_api_v1_pa_payment_intents__id__cancel/post)
-  [Webhooks](https://www.airwallex.com/docs/developer-tools__listen-for-webhook-events)

### Additional Changes

- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

The Refund test cases were ignored as the implementation has a external permission blocker from the connector.

Screenshot 2023-02-23 at 6 21 04 PM

## Checklist

- [X] I formatted the code `cargo +nightly fmt --all`
- [X] I addressed lints thrown by `cargo clippy`
- [X] I reviewed submitted code
- [X] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

**Review by jagan-jaya — DISMISSED** [review]

**Review by ArjunKarthik — DISMISSED** [review]

**Review by jagan-jaya — APPROVED** [review]

**Review by ArjunKarthik — DISMISSED** [review]

File: crates/router/src/connector/airwallex/transformers.rs [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+
+#[derive(Debug, Serialize, Eq, PartialEq)]
+pub struct AirwallexCard {
+    card: AirwallexCardDetails,
+    #[serde(rename = "type")]
+    payment_method_type: AirwallexPaymentType,
+}
+#[derive(Debug, Serialize, Eq, PartialEq)]
+pub struct AirwallexCardDetails {
+    expiry_month: String,
```
Comment (jarnura) [inline crates/router/src/connector/airwallex/transformers.rs]: Make `secrets` for  `expiry_month`, `expiry_year`, `cvc`

File: crates/router/src/connector/airwallex/transformers.rs [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+impl TryFrom<&types::PaymentsAuthorizeSessionTokenRouterData> for AirwallexIntentRequest {
+    type Error = error_stack::Report<errors::ConnectorError>;
+    fn try_from(
+        item: &types::PaymentsAuthorizeSessionTokenRouterData,
+    ) -> Result<Self, Self::Error> {
+        Ok(Self {
+            request_id: Uuid::new_v4().to_string(),
+            amount: item.request.amount,
+            currency: item.request.currency,
+            merchant_order_id: Uuid::new_v4().to_string(),
```
Comment (jarnura) [inline crates/router/src/connector/airwallex/transformers.rs]: What this `merchant_order_id` refers to?

File: crates/router/src/connector/airwallex/transformers.rs [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+                    },
+                    payment_method_type: AirwallexPaymentType::Card,
+                }))
+            }
+            _ => Err(errors::ConnectorError::NotImplemented(
+                "Unknown payment method".to_string(),
+            )),
+        }?;
+        Ok(Self {
+            request_id: Uuid::new_v4().to_string(),
```
Comment (jarnura) [inline crates/router/src/connector/airwallex/transformers.rs:106]: What this `request_id` refers to?

File: crates/router/src/connector/airwallex/transformers.rs [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+    currency: enums::Currency,
+    merchant_order_id: String,
+}
+impl TryFrom<&types::PaymentsAuthorizeSessionTokenRouterData> for AirwallexIntentRequest {
+    type Error = error_stack::Report<errors::ConnectorError>;
+    fn try_from(
+        item: &types::PaymentsAuthorizeSessionTokenRouterData,
+    ) -> Result<Self, Self::Error> {
+        Ok(Self {
+            request_id: Uuid::new_v4().to_string(),
```
Comment (jarnura) [inline crates/router/src/connector/airwallex/transformers.rs:26]: What this `request_id` refers to?

File: crates/router/src/connector/airwallex/transformers.rs (hunk 2 above)
Comment (jagan-jaya) [inline crates/router/src/connector/airwallex/transformers.rs:30]: change it to pament_intent_id

File: crates/router/src/connector/airwallex/transformers.rs [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+                            item.request.capture_method,
+                            Some(enums::CaptureMethod::Automatic) | None
+                        ),
+                    }));
+                Ok(AirwallexPaymentMethod::Card(AirwallexCard {
+                    card: AirwallexCardDetails {
+                        number: ccard.card_number,
+                        expiry_month: ccard.card_exp_month.peek().to_string(),
+                        expiry_year: ccard.card_exp_year.peek().to_string(),
+                        cvc: ccard.card_cvc.peek().to_string(),
```
Comment (jagan-jaya) [inline crates/router/src/connector/airwallex/transformers.rs:97]: do not peek into pii info, change it to secret

**Review by jagan-jaya — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/airwallex/transformers.rs [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+use crate::{
+    core::errors,
+    pii::{self, Secret},
+    services,
+    types::{self, api, storage::enums},
+};
+
+#[derive(Default, Debug, Serialize, Eq, PartialEq)]
+pub struct AirwallexIntentRequest {
+    request_id: String,
```
Comment (jarnura) [inline crates/router/src/connector/airwallex/transformers.rs:15]: Add comments at-least for id's, what is that id and why it is used.

Comment (jarnura) [reply crates/router/src/connector/airwallex/transformers.rs:15]: Do the same for all other request and response structs also

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/airwallex/transformers.rs (hunk 2 above)
Comment (SamraatBansal) [reply crates/router/src/connector/airwallex/transformers.rs]: It is ID created in merchant's order system that corresponds to this PaymentIntent.
Already changed to `payment_id`

File: crates/router/src/connector/airwallex/transformers.rs (hunk 3 above)
Comment (SamraatBansal) [reply crates/router/src/connector/airwallex/transformers.rs:106]: Unique ID to be sent for each transaction/operation request to the connector

File: crates/router/src/connector/airwallex/transformers.rs (hunk 4 above)
Comment (SamraatBansal) [reply crates/router/src/connector/airwallex/transformers.rs:26]: Unique ID to be sent for each transaction/operation request to the connector

File: crates/router/src/connector/airwallex/transformers.rs (hunk 6 above)
Comment (SamraatBansal) [reply crates/router/src/connector/airwallex/transformers.rs:16]: Done.

File: crates/router/src/connector/airwallex.rs [hunk 7]
```diff
@@ -0,0 +1,916 @@
...
+impl api::PaymentAuthorize for Airwallex {}
+
+#[async_trait::async_trait]
+impl ConnectorIntegration<api::Authorize, types::PaymentsAuthorizeData, types::PaymentsResponseData>
+    for Airwallex
+{
+    async fn execute_pretasks(
+        &self,
+        router_data: &mut types::PaymentsAuthorizeRouterData,
+        app_state: &crate::routes::AppState,
```
Comment (jarnura) [inline crates/router/src/connector/airwallex.rs:287]: Remove absolute path for imports

File: crates/router/src/connector/airwallex.rs [hunk 8]
```diff
@@ -0,0 +1,916 @@
...
+    fn get_content_type(&self) -> &'static str {
+        self.common_get_content_type()
+    }
+
+    fn get_url(
+        &self,
+        req: &types::RefundSyncRouterData,
+        connectors: &settings::Connectors,
+    ) -> CustomResult<String, errors::ConnectorError> {
+        let refund_id = super::utils::RefundsRequestData::get_connector_refund_id(&req.request)?;
```
Comment (jarnura) [inline crates/router/src/connector/airwallex.rs:751]: Remove absolute path for imports

File: crates/router/src/connector/airwallex.rs [hunk 9]
```diff
@@ -0,0 +1,916 @@
...
+            })
+            .ok_or(errors::ConnectorError::WebhookSignatureNotFound)
+            .into_report()??;
+
+        Ok(format!("{}{}", timestamp, String::from_utf8_lossy(request.body)).into_bytes())
+    }
+
+    async fn get_webhook_source_verification_merchant_secret(
+        &self,
+        db: &dyn crate::db::StorageInterface,
```
Comment (jarnura) [inline crates/router/src/connector/airwallex.rs:857]: Remove absolute path for imports

**Review by jarnura — CHANGES_REQUESTED** [review]

**Review by jagan-jaya — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

**Pull request merged**